databricks/koalas/plot/matplotlib.py

```diff
@@ -566,6 +577,142 @@ def _compute_hist(sdf, bins):
 
         return pdf[bucket_name]
 
+    @staticmethod
+    def _compute_hist_multi_columns(sdf, bins):
+        """
+        It can calculate histogram with multiple columns in single pass. `_compute_hist_v1` can
+        only be able to do it each job for each column.
+        """
+
+        # 'data' is a Spark DataFrame that selects one column.
+        assert isinstance(bins, (np.ndarray, np.generic))
+
+        colnames = sdf.columns
+
+        bucket_names = ["__{}_bucket".format(colname) for colname in colnames]
+        # creates a Bucketizer to get corresponding bin of each value
+        bucketizer = Bucketizer(
+            splitsArray=[bins] * len(colnames),
+            inputCols=colnames,
+            outputCols=bucket_names,
+            handleInvalid="skip",
+        )
+
+        # 1. Make the bucket output flat. For example from
+        #     +-------+-------+-----------------+-----------------+
+        #     |values1|values2|__values1__bucket|__values2__bucket|
+        #     +-------+-------+-----------------+-----------------+
+        #     |0.1    |0.0    |0.0              |0.0              |
+        #     |0.4    |1.0    |0.0              |1.0              |
+        #     |1.2    |1.3    |1.0              |1.0              |
+        #     |1.5    |NaN    |2.0              |2.0              |
+        #     |NaN    |1.0    |3.0              |1.0              |
+        #     |NaN    |0.0    |3.0              |0.0              |
+        #     +-------+-------+-----------------+-----------------+
+        #
+        # to:
+        #
+        #     +----------+-------+
+        #     |__group_id|buckets|
+        #     +----------+-------+
+        #     |0         |0.0    |
+        #     |0         |0.0    |
+        #     |0         |1.0    |
+        #     |0         |2.0    |
+        #     |0         |3.0    |
+        #     |0         |3.0    |
+        #     |1         |0.0    |
+        #     |1         |1.0    |
+        #     |1         |1.0    |
+        #     |1         |2.0    |
+        #     |1         |1.0    |
+        #     |1         |0.0    |
+        #     +----------+-------+
+        bucket_df = bucketizer.transform(sdf)
+        output_df = None
+
+        for group_id, bucket_name in enumerate(bucket_names):
+            if output_df is None:
+                output_df = bucket_df.select(
+                    F.lit(group_id).alias("__group_id"), F.col(bucket_name).alias("__bucket")
+                )
+            else:
+                output_df = output_df.union(
+                    bucket_df.select(
+                        F.lit(group_id).alias("__group_id"), F.col(bucket_name).alias("__bucket")
+                    )
+                )
```

Comment:
If we try to flatten the output like this, we can just union single-column transformations for each column? Then we can apply it to older Spark?

Reply:
Oh hm, yeah I think we can. Let me try to combine both.